include/cufinufft_eitherprec.h

```diff
@@ -0,0 +1,299 @@
+#define CUINTERP3D cuinterp3df
+#define CUINTERP2D_NUPTSDRIVEN cuinterp2df_nuptsdriven
+#define CUINTERP2D_SUBPROB cuinterp2df_subprob
+#define CUINTERP3D_NUPTSDRIVEN cuinterp3df_nuptsdriven
+#define CUINTERP3D_SUBPROB cuinterp3df_subprob
+/* deconvolve */
+#define CUDECONVOLVE2D cudeconvolve2df
+#define CUDECONVOLVE3D cudeconvolve3df
+/* structs */
+#define CUFINUFFT_OPTS cufinufftf_opts
+#define CUFINUFFT_PLAN cufinufftf_plan
+
+#else
+
+#define CUFINUFFT_DEFAULT_OPTS cufinufft_default_opts
+#define CUFINUFFT_MAKEPLAN cufinufft_makeplan
+#define CUFINUFFT_SETNUPTS cufinufft_setNUpts
+#define CUFINUFFT_EXEC cufinufft_exec
+#define CUFINUFFT_DESTROY cufinufft_destroy
+#define CUFINUFFT2D1_EXEC cufinufft2d1_exec
+#define CUFINUFFT2D2_EXEC cufinufft2d2_exec
+#define CUFINUFFT3D1_EXEC cufinufft3d1_exec
+#define CUFINUFFT3D2_EXEC cufinufft3d2_exec
+#define SETUP_BINSIZE setup_binsize
+/* extern c interface */
+#define CUFINUFFTC_DEFAULT_OPTS cufinufftc_default_opts
+#define CUFINUFFTC_MAKEPLAN cufinufftc_makeplan
+#define CUFINUFFTC_SETNUPTS cufinufftc_setNUpts
+#define CUFINUFFTC_EXEC cufinufftc_exec
+#define CUFINUFFTC_DESTROY cufinufftc_destroy
+/* memtransfer.h */
+#define ALLOCGPUMEM1D_PLAN allocgpumem1d_plan
+#define ALLOCGPUMEM1D_NUPTS allocgpumem1d_nupts
+#define FREEGPUMEMORY1D freegpumemory1d
+#define ALLOCGPUMEM2D_PLAN allocgpumem2d_plan
+#define ALLOCGPUMEM2D_NUPTS allocgpumem2d_nupts
+#define FREEGPUMEMORY2D freegpumemory2d
+#define ALLOCGPUMEM3D_PLAN allocgpumem3d_plan
+#define ALLOCGPUMEM3D_NUPTS allocgpumem3d_nupts
+#define FREEGPUMEMORY3D freegpumemory3d
+/* spreading 2D */
+#define CUSPREAD2D cuspread2d
+#define CUSPREAD2D_NUPTSDRIVEN_PROP cuspread2d_nuptsdriven_prop
+#define CUSPREAD2D_NUPTSDRIVEN cuspread2d_nuptsdriven
+#define CUSPREAD2D_SUBPROB_PROP cuspread2d_subprob_prop
+#define CUSPREAD2D_SUBPROB cuspread2d_subprob
+#define CUSPREAD2D_PAUL cuspread2d_paul
+#define CUSPREAD2D_PAUL_PROP cuspread2d_paul_prop
+/* spreading 3d */
+#define CUSPREAD3D cuspread3d
+#define CUSPREAD3D_NUPTSDRIVEN_PROP cuspread3d_nuptsdriven_prop
+#define CUSPREAD3D_NUPTSDRIVEN cuspread3d_nuptsdriven
+#define CUSPREAD3D_BLOCKGATHER_PROP cuspread3d_blockgather_prop
+#define CUSPREAD3D_BLOCKGATHER cuspread3d_blockgather
+#define CUSPREAD3D_SUBPROB_PROP cuspread3d_subprob_prop
+#define CUSPREAD3D_SUBPROB cuspread3d_subprob
+/* interp */
+#define CUINTERP2D cuinterp2d
+#define CUINTERP3D cuinterp3d
+#define CUINTERP2D_NUPTSDRIVEN cuinterp2d_nuptsdriven
+#define CUINTERP2D_SUBPROB cuinterp2d_subprob
+#define CUINTERP3D_NUPTSDRIVEN cuinterp3d_nuptsdriven
+#define CUINTERP3D_SUBPROB cuinterp3d_subprob
+/* deconvolve */
+#define CUDECONVOLVE2D cudeconvolve2d
+#define CUDECONVOLVE3D cudeconvolve3d
+/* structs */
+#define CUFINUFFT_OPTS cufinufft_opts
+#define CUFINUFFT_PLAN cufinufft_plan
+
+#endif
+
+typedef struct CUFINUFFT_OPTS {   // see cufinufft_default_opts() for defaults
+	FLT upsampfac;   // upsampling ratio sigma, only 2.0 (standard) is implemented
```

Comment:
From a code coherence perspective, mixing floats and doubles when convenient is probably worse. However, that's just my opinion, so I've made the change here so it is consistent with what you're suggesting from FINUFFT (since having them be similar as possible was one of the main goals).

I checked and I do not believe it is used directly in any kernels (thus keeping them in totally singles or doubles).  I would have stronger opinions about that...

Reply:
Hello. fixing upsampfac as double is what I did in finufft, for the reason J gave. I think it's simpler for interface design. It doesn't affect the kernel eval, because it's really just a "flag" whose value (2.0, 1.25) triggers different kernel formulae.

For reference here's the relevant finufft commits:
https://github.com/flatironinstitute/finufft/commit/db18e4f82e9642d6ec98de57dc7c273fe8858ab3
https://github.com/flatironinstitute/finufft/commit/06311f44804925cb12212600190bb878025dd2b8